django/http/request.py

```diff
@@ -369,6 +369,14 @@ def __init__(self, environ):
                 headers[name] = value
         super().__init__(headers)
 
+    def __getitem__(self, key):
+        """
+        Django template variable lookup precludes hyphens. By converting
+        underscores to hyphens here, headers['foo-bar'] can be used
+        in views, and {{ headers.foo_bar }} in templates.
```

Comment:
`headers` should be `request.headers` in both cases, but I'd be inclined to simplify this:

```python
"""Allows header lookup using underscores in place of hyphens."""
```